src/lib/util.ts

```diff
@@ -66,3 +67,121 @@ export const formatParamValueForDisplay = (value: number | string) => {
+	[MIDIMetaMappingType.KeyPressure]: /^KPRESS#(?<keypress>[0-9]{1,3})\/(?<chan>[0-9]{1,2})$/,
+	[MIDIMetaMappingType.Note]: /^NOTE#(?<note>[0-9]{1,3})\/(?<chan>[0-9]{1,2})$/,
+	[MIDIMetaMappingType.PitchBend]: /^BEND\/(?<bend>[0-9]{1,2})$/,
+	[MIDIMetaMappingType.ProgramChange]: /^PRGCHG\/(?<prgchg>[0-9]{1,2})$/
+};
+
+export const parseMIDIMappingDisplayValue = (value: string): false | { type: MIDIMetaMappingType, mapping: MIDIMetaMapping } => {
+	for (const [mappingType, reg] of Object.entries(midiMetaRegexp) as Array<[MIDIMetaMappingType, RegExp]>) {
+		const match = value.match(reg);
+		if (!match) continue;
+
+		switch (mappingType) {
+			case MIDIMetaMappingType.ChannelPressure: {
+				const chanpress = parseInt(match.groups.chanpress, 10);
+				if (isNaN(chanpress) || chanpress < 1 || chanpress > 16) return false;
+				return {
+					type: MIDIMetaMappingType.ChannelPressure,
+					mapping: { chanpress } as MIDIChannelPressureMetaMapping
+				};
+			}
+			case MIDIMetaMappingType.ControlChange: {
+				const chan = parseInt(match.groups.chan, 10);
+				if (isNaN(chan) || chan < 1 || chan > 16) return false;
+
+				const ctrl = parseInt(match.groups.ctrl, 10);
+				if (isNaN(ctrl) || ctrl < 0 || ctrl > 127) return false;
+
+				return {
+					type: MIDIMetaMappingType.ControlChange,
+					mapping: { chan, ctrl } as MIDIControlChangeMetaMapping
+				};
+			}
+			case MIDIMetaMappingType.KeyPressure: {
+				const chan = parseInt(match.groups.chan, 10);
+				if (isNaN(chan) || chan < 1 || chan > 16) return false;
+
+				const keypress = parseInt(match.groups.keypress, 10);
+				if (isNaN(keypress) || keypress < 0 || keypress > 127) return false;
+
+				return {
+					type: MIDIMetaMappingType.KeyPressure,
+					mapping: { chan, keypress } as MIDIKeypressMetaMapping
+				};
+
+			}
+			case MIDIMetaMappingType.Note: {
+				const chan = parseInt(match.groups.chan, 10);
+				if (isNaN(chan) || chan < 1 || chan > 16) return false;
+
+				const note = parseInt(match.groups.note, 10);
+				if (isNaN(note) || note < 0 || note > 127) return false;
+
+				return {
+					type: MIDIMetaMappingType.Note,
+					mapping: { chan, note } as MIDINoteMetaMapping
+				};
+
+			}
+			case MIDIMetaMappingType.PitchBend: {
+				const bend = parseInt(match.groups.bend, 10);
+				if (isNaN(bend) || bend < 1 || bend > 16) return false;
+				return {
+					type: MIDIMetaMappingType.PitchBend,
+					mapping: { bend } as MIDIPitchBendMetaMapping
+				};
+
+			}
+			case MIDIMetaMappingType.ProgramChange: {
+				const prgchg = parseInt(match.groups.prgchg, 10);
+				if (isNaN(prgchg) || prgchg < 1 || prgchg > 16) return false;
+				return {
+					type: MIDIMetaMappingType.ProgramChange,
+					mapping: { prgchg } as MIDIProgramChangeMetaMapping
+				};
+			}
+
+			default: {
+				// eslint-disable-next-line @typescript-eslint/no-unused-vars
+				const _exhaustive: never = mappingType;
+				throw new Error(`Unknown MIDIMappingType "${mappingType}"`);
```

Comment:
this throwing seems okay because it is from a type we've defined here? so it would be an error to not be able to render a known type?

Reply:
this does basically two things - ensure that the switch / case is exhaustive with the `never` type and then in order to have some form of "return" value it throws. But the function is used to parse user input from the MIDI source field to a MIDI mapping.